docs/source/csdl-spec.md

```diff
@@ -0,0 +1,630 @@
+
+The CSDL is intended to be a single artifact which can power a graph router. For example, the CSDL replaces the service list configuration of Apollo Gateway.
+
+CSDL is a subset of the [GraphQL Schema Definition Language](https://spec.graphql.org/). It removes certain SDL features (for instance, type extensions are not permitted) while also requiring the definition of certain types and providing a suite of directives to describe graph topology. In particular, a CSDL can:
+- define [subgraphs](#subgraph) and bind them to [endpoints](#endpoint) (with the required [`csdl_Graph`](#csdl_Graph) type and [`@csdl_endpoint`](#csdl_endpoint) directive)
+- assign fields to subgraphs (with [`@csdl_resolve`](#resolve))
+- declare additional data required and provided by subgraph field resolvers (with [`@csdl_key`](#key), and [`@csdl_resolve`](#csdl_resolve))
+
+## How to read this document
+
+TK, discussion of spec nomenclature
+e.g.: https://html.spec.whatwg.org/multipage/introduction.html#how-to-read-this-specification
+
+### What this document isn't
+
+This document specifies the CSDL and only the CSDL. It does not have an opinion about how CSDL should be generated from subgraphs. A suggestion is provided in [Appendix: Basic Composition Algorithm](#appendix-suggested-composition-algorithm), but conforming implementations may choose approach they like.
+
+## Example: Let's go to the moon
+
+*This section is non-normative.*
+
+We'll refer to this example throughout the document. The example consists of two example schemas—a `rockets` schema which serves data about rockets, and an `astronauts` service which services queries regarding the people flying on them—along with a CSDL that composes them.
+
+### Rockets
+
+<a name=ex-rockets href=#ex-rockets class='listing not-csdl'>Rockets example service</a>
+```graphql
+# Rockets Service
+type Query {
+  rockets: [Rocket]!
+}
+
+type Rocket @key(fields: "id") {
+  id: String! 
+  name: String!
+  captain: Astronaut @provides(fields: "tripId")
+}
+
+extend type Astronaut @key(fields: "id") {
+  id: String! @external
+  tripId: String! @external
+  rocket: Rocket!
+}
+```
+
+### Astronauts
+
+<a name=ex-astronauts href=#ex-astronauts class='listing not-csdl'>Astronauts example service</a>
+
+```graphql
+# Astronauts Service
+type Query {
+  astronauts: [Astronaut]!
+}
+
+type Astronaut @key(fields: "id") {
+  id: String! 
+  name: String!
+  tripId: String!
+}
+
+extend type Rocket @key(fields: "id") {
+  id: String! @external
+  astronaut: Astronaut!
+}
+```
+
+### CSDL: Rockets & Astronauts
+
+This example is non-normative, and represents just one possible way to compose [rockets](#rockets) and [astronauts](#astronauts) into CSDL.
+
+<a name=ex-rockets-and-astronauts href=#ex-rockets-and-astronauts class=listing>Example CSDL composing `rockets` and `astronauts`</a>
+```graphql
+schema @csdl(version: "^1") {
+  query: Query
+}
+
+enum csdl_Graph {
```

Comment:
This is a surprising choice on first pass to use nested directives and enums to indicate the set of sub-graphs. What are the advantages of this approach vs. the existing parameterized directive `@csdl_graph` approach?

Reply:
Ahh, it's because we can refer to that graph with all of its required metadata repeatably. That makes a lot of sense! I should have kept scrolling 😉 May I suggest that we use `@csdl_graph_meta` or something more generic? I suspect that we may want to have information like `revision: 1234` represented in the CSDL, and things like fall-back strategies that are at the sub-graph level down the road.

Reply:
I think if we want to attach more information to an endpoint, we can use different / more directives?

Is there anything that must be represented—i.e. is there endpoint-specific config that gateways always need which we should consider either adding to `@csdl_endpoint` or introducing new directives for?